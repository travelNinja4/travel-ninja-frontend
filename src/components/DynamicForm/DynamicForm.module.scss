/**
------------------------------------------------------
styles specific to the DynamicForm component
------------------------------------------------------
**/

@use '../../styles/mixins' as *;

.field {
  display: flex;
  flex-direction: column;
  flex: 0 0 100%;
  min-width: 0;
  width: 100%;
}

.field--full {
  flex: 0 0 100%;
  max-width: 100%;
}

.field--half {
  flex: 0 0 100%;
}

.field--third {
  flex: 0 0 100%;
}

/* ðŸ‘‰ At tablet size */
@include respond-to(md) {
  .field--half {
    flex: 0 0 calc(50% - 0.5rem);
    max-width: calc(50% - 0.5rem);
  }

  .field--third {
    flex: 0 0 calc((100% - 2rem) / 3);
    max-width: calc((100% - 2rem) / 3);
  }
}

/* ðŸ‘‰ At desktop size */
@include respond-to(lg) {
  .field--half {
    flex: 1 1 calc(50% - 1rem);
    max-width: calc(50% - 1rem);
  }

  .field--third {
    flex: 1 1 calc(33.333% - 1rem);
    max-width: calc(33.333% - 1rem);
  }
}
