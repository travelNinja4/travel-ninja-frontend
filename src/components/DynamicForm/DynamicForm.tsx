/**
 * The DynamicForm component is a reusable dynamic form builder that generates form fields based on a given configuration
 *
 * @example
 * ```tsx
 * import DynamicForm from '@src/components/DynamicForm'
 *
 * export default function DynamicForm() {
 *   return <DynamicForm label="Hello" />;
 * }
 * ```
 */
'use client';

import React, { ReactNode } from 'react';
import { useForm, SubmitHandler, FieldValues, Resolver, Path, Controller } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';
import { ZodSchema } from 'zod';
import Typography from '../Typography';
import { Tags, TextAlign } from '../Typography/Typography';
import TextField from '../TextField';
import PasswordFeedback from '../PasswordFeedback';
import { PasswordHint } from '../PasswordFeedback/PasswordFeedback';
import CheckBox from '../CheckBox';
import styles from './DynamicForm.module.scss';
import Button from '../Button';
import { LucideIcon } from 'lucide-react';

/**
 * Define the props available for the DynamicForm component.
 */
interface DynamicFormProps<T extends FieldValues = FieldValues> {
  fields: FieldConfig[];
  schema: ZodSchema<T>;
  className: string;
  onSubmit: SubmitHandler<T>;
}

/**
 * Enum for safe width values
 */
export enum FieldWidth {
  FULL = 'full',
  HALF = 'half',
  THIRD = 'third',
}

export interface FieldConfig {
  name: string;
  label?: string | ReactNode;
  type?: FieldType;
  placeholder?: string;
  tag?: Tags;
  className?: string;
  children?: ReactNode;
  width?: FieldWidth;
  required?: boolean;
  validationHints?: PasswordHint[];
  buttonType?: 'button' | 'submit' | 'reset';
  startIcon?: LucideIcon;
  endIcon?: LucideIcon;
  iconColor?: string;
  textAlign?: TextAlign;
}

type FieldType =
  | 'label'
  | 'input'
  | 'email'
  | 'tel'
  | 'password'
  | 'textarea'
  | 'checkbox'
  | 'select'
  | 'button';

export default function DynamicForm<T extends FieldValues>({
  fields,
  schema,
  className,
  onSubmit,
}: DynamicFormProps<T>) {
  const {
    register,
    handleSubmit,
    watch,
    control,
    formState: { errors },
  } = useForm<T>({
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    resolver: zodResolver(schema as any) as unknown as Resolver<T>,
    mode: 'all',
    reValidateMode: 'onChange',
  });

  const renderField = (field: FieldConfig) => {
    const value = watch(field.name as Path<T>) || '';
    const commonProps = {
      ...field,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      ...register(field.name as any),
      error: errors[field.name]?.message as string,
    };
    switch (field.type) {
      case 'label':
        return (
          <Typography tag={field.tag || 'p'} className={field.className} align={field.textAlign}>
            {field.label || field.children}
          </Typography>
        );

      case 'input':
        return <TextField {...commonProps} className={field.className} />;

      case 'password':
        return (
          <>
            <TextField {...commonProps} className={field.className} />
            <PasswordFeedback value={value} hints={field.validationHints ?? []} />
          </>
        );

      case 'checkbox':
        return (
          <Controller
            name={field.name as Path<T>}
            control={control}
            render={({ field: { onChange, value } }) => (
              <CheckBox
                label={field.label as ReactNode}
                checked={!!value}
                onChange={(checked: boolean) => onChange(checked)}
                className={field.className}
              />
            )}
          />
        );

      case 'button':
        return (
          <Button
            key={field.name}
            type={field.buttonType || 'submit'}
            className={field.className}
            startIcon={field.startIcon}
            endIcon={field.endIcon}
            iconColor={field.iconColor}
          >
            {field.label}
          </Button>
        );

      default:
        null;
    }
  };

  return (
    <form data-testid="DynamicFormTest" className={className} onSubmit={handleSubmit(onSubmit)}>
      {fields.map((field) => (
        <div
          key={field.name}
          className={`${className}__field ${className}__field--${field.width || FieldWidth.FULL}`}
        >
          {renderField(field)}
        </div>
      ))}
    </form>
  );
}
