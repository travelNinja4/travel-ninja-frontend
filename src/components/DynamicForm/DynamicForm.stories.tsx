import type { Meta, StoryObj } from '@storybook/nextjs-vite';
/**
import { fn } from 'storybook/test';
*/
import DynamicForm from './DynamicForm';

const meta = {
  title: 'components/DynamicForm',
  component: DynamicForm,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {},
  args: {
    /** onClick: fn() **/
  },
} satisfies Meta<typeof DynamicForm>;

export default meta;
type Story = StoryObj<typeof meta>;
