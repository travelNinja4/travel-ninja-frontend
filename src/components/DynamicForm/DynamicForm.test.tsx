import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { z } from 'zod';
import DynamicForm, { FieldConfig, FieldWidth } from './DynamicForm';

const mockFields: FieldConfig[] = [
  {
    name: 'username',
    label: 'Username',
    type: 'input',
    placeholder: 'Enter username',
    width: FieldWidth.FULL,
  },
  {
    name: 'password',
    label: 'Password',
    type: 'password',
    placeholder: 'Enter password',
    validationHints: [],
  },
];

const mockSchema = z.object({
  username: z.string().min(1, 'Username is required'),
  password: z.string().min(6, 'Password must be at least 6 characters'),
});

describe('Home', () => {
  /** Base props for DynamicForm **/
  const baseProps = {
    fields: mockFields,
    schema: mockSchema,
    className: 'dynamic-form',
    onSubmit: jest.fn(),
  };

  /**
   * @test
   * Renders the DynamicForm component without throwing an error
   * and checks if the DynamicFormTest text is present in the document.
   *
   * This ensures that the component loads correctly with the base props.
   */
  it('renders without crashing', () => {
    render(<DynamicForm {...baseProps} />);
    expect(screen.getByTestId('DynamicFormTest')).toBeInTheDocument();
  });

  /**
   * @test
   * Ensures that the DynamicForm component renders without throwing any runtime errors.
   *
   * This is a safety check to confirm that the component can mount successfully
   * with the provided base props, even if it does not render any visible content.
   */
  it('renders without throwing', () => {
    expect(() => render(<DynamicForm {...baseProps} />)).not.toThrow();
  });

  /**
   * @test
   * Captures a snapshot of the DynamicForm component with base props
   * and ensures it matches the previously stored snapshot.
   *
   * This helps detect any unexpected UI changes.
   */
  it('matches snapshot', () => {
    const { asFragment } = render(<DynamicForm {...baseProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
