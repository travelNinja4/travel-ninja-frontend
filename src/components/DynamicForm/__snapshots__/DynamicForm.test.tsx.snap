// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Home matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="dynamic-form"
    data-testid="DynamicFormTest"
  >
    <div
      class="field field--full"
    >
      <div
        class="container"
        data-testid="TextFieldTest"
      >
        <label
          class="typography label align-left label"
          data-testid="TypographyTest"
        >
          Username 
        </label>
        <div
          class="inputWrapper"
        >
          <input
            class="inputField"
            name="username"
            placeholder="Enter username"
            type="input"
            width="full"
          />
        </div>
      </div>
    </div>
    <div
      class="field field--full"
    >
      <div
        class="container"
        data-testid="TextFieldTest"
      >
        <label
          class="typography label align-left label"
          data-testid="TypographyTest"
        >
          Password 
        </label>
        <div
          class="inputWrapper"
        >
          <input
            class="inputField inputWithIcon"
            name="password"
            placeholder="Enter password"
            type="password"
            validationhints=""
          />
          <div
            class="iconButton"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-eye"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M2.062 12.348a1 1 0 0 1 0-.696 10.75 10.75 0 0 1 19.876 0 1 1 0 0 1 0 .696 10.75 10.75 0 0 1-19.876 0"
              />
              <circle
                cx="12"
                cy="12"
                r="3"
              />
            </svg>
          </div>
        </div>
      </div>
      <ul
        aria-live="polite"
        class="container"
        data-testid="PasswordFeedbackTest"
      />
    </div>
  </form>
</DocumentFragment>
`;
